clusterloader2/api/types.go

```diff
@@ -147,8 +147,8 @@ type TuningSet struct {
 type Measurement struct {
 	// Method is a name of a method registered in the ClusterLoader factory.
 	Method string `json: method`
-	// Identifier is a string that differentiates measurement instances of the same method.
-	Identifier string `json: identifier`
+	// Identifier is an interface{} which could be a string or map of {name: value} that differentiates measurement instances of the same method.
```

Comment:
Do we support anything else than:
(a) string
(b) map[string]string
?

If not, I would prefer to make it
```
type Identifier struct {
  // ... exactly one of those should be set
  Identifier string
  SomeNiceName map[string]string
}
```

I don't think we really need multi-nesting, so that would be easier to understand.

Reply:
@wojtek-t seems to be that the structure can be a bit more nested than just a map of strings.
what we can have is identifier specific params. For example:

```
Identifier:
  A:
    Param1: X
    Param2: X'
    Param3; X''
  B:
    Param1: Y
    Param2: Y'
    Param3; Y''
```

what it means is that the values of the map with keys A and B are map of strings. plz let me know in case i am missing something.

Reply:
So ma[string]map[string]string.
Can it be different than string and that?

Reply:
yeah I think that's the max we can go for these params. They should have the same structures to what params have today, just that they are nested within service Identifier to signify identifier specific overrides to the outer params.
Unless @mm4tt has some other ideas?

Reply:
So actually this can't even be map[string]string, right?
It can only be either:
1. string
2. map[string]map[string]string

So I think my initial proposal holds (just the types are slightly different).

That said, looking into the API pkg:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/api/types.go#L153
we already have map[string]interface{} there (which I don't really like).
But we definitely should be consistent with it - so let's stick to map[string]interface{} for now.


So basically, let's do the following:
```
type MeasurementConfig struct {
  Identifier string
  Params map[string]interface{}
}

type Measurement struct {
  // method as is

  // Exactly one of this should be filled in:
  Identifier & Params as is

  Configs []MeasurementConfig
}
```

Once this is merged, we should also get rid of Identifier & Params and always used the Configs slice.

[I'm open to any better names suggestions].

@vamossagar12 @mm4tt - thoughts?

Reply:
hey @wojtek-t  MeasurementConfig already exists in the measurement package so using the same name can be a bit confusing? Perhaps we can call it IdentifierConfig?

Also, since it effectively is moving 2 fields to a separate struct, what advantages would we gain this way? To my understanding, this Identifier specific config is internal to a  measurement so it's not that we can make use of this config from outside.

Reply:
> hey @wojtek-t MeasurementConfig already exists in the measurement package so using the same name can be a bit confusing? Perhaps we can call it IdentifierConfig?

That's not a problem - names have to be unique only within package.

> Also, since it effectively is moving 2 fields to a separate struct, what advantages would we gain this way? To my understanding, this Identifier specific config is internal to a measurement so it's not that we can make use of this config from outside.

It's explicitly bundling two things that should be bundled - params doesn't make sense without identifier.

Reply:
@wojtek-t Ok got it. Would separate the MeasurementConfig out. 1 last thing though:

The way I had implemented the earlier change, based upon the type of the Identifier I would invoke either the pre-existing singular measurement(if it's a string) or the newly created wrapper measurement(if it's a map which means it can contain more than 1 measurements). 
Example: 
https://github.com/kubernetes/perf-tests/pull/1075/files#diff-1f56f857f166b4d383af27f485c038daR30

That nested map would internally have more params within it which is specific to the measurement. 

Example config:
nested identifier:
https://github.com/kubernetes/perf-tests/pull/1075/files#diff-ead315efdb78ab25dc9df57a1f86f356R24

vs

https://github.com/kubernetes/perf-tests/pull/1075/files#diff-ead315efdb78ab25dc9df57a1f86f356R33

The above example has some more params with the measurement name as the key.

The outer Params map still remains as is. as can be found here:

https://github.com/kubernetes/perf-tests/pull/1075/files#diff-ead315efdb78ab25dc9df57a1f86f356R31

With the new structure proposed by you, Identifier is a string and Params has been moved out but there are 2 params which can be supplied not in the case of wrapped measurements: 1) Measurement specific and 2) outer which are applicable to all the measurements within the wrapper measurements.

I know there has been a lot of to and fros from me on this particular item(apologies for that!) but I am hoping with this last question I should be able to close it out. Thanks!

Reply:
> With the new structure proposed by you, Identifier is a string and Params has been moved out but there are 2 params which can be supplied not in the case of wrapped measurements: 1) Measurement specific and 2) outer which are applicable to all the measurements within the wrapper measurements.

I think I didn't fully understand it. So let me clarify based on my understanding and let me know if it's not what you meant.

What I'm suggesting currrently is the following:

```
type MeasurementConfig struct {
  Identifier string
  Params map[string]interface{}
}

type Measurement struct {
  // method as is

  // Identifier as is
  Identifier string
  // Params as is
  map[string]interface{}

  Configs []MeasurementConfig
}
```

We should also introduce a check that exactly one of `Identifier` and `Configs` should be non-empty.

That said I agree, that we may want to allow Params can still be set, even if `Configs` is non-empty. In that case, params for individual measurements should be contatenation of the specific + the common ones.

So in fact how it conceptually will look like is:
```
type Measurement struct {
  // method as is

  // params as is

  // Exactly one of the below fields should be non-empty.

  // Identifier as is
  Identifier string

  Configs []MeasurementConfig
}
```

Does that answer your question? Let me know if anything is still unclear.

Reply:
So, to summarize. What we've had before in this PR was:
```
- Method: ServiceCreationLatency
    Identifier:
      ServiceCreationLatencyLarge:
       labelSelector: size = {{$LARGE_BACKEND_LABEL}}
      ServiceCreationLatencyMedium:
       labelSelector: size = {{$MEDIUM_BACKEND_LABEL}}
      ServiceCreationLatencySmall:
       labelSelector: size = {{$SMALL_BACKEND_LABEL}}
    Params:
     action: start
```

What Wojtek is proposing would be:
```
- Method: ServiceCreationLatency
   Configs:
      - Identifier: ServiceCreationLatencyLarge
        Params:
          labelSelector: size = {{$LARGE_BACKEND_LABEL}}
      - Identifier: ServiceCreationLatencyMedium
        Params:
          labelSelector: size = {{$MEDIUM_BACKEND_LABEL}}
      - Identifier: ServiceCreationLatencySmall
        Params:
          labelSelector: size = {{$SMALL_BACKEND_LABEL}}
    Params:
     action: start
```

I think it's fine. It has a bit more boilerplate than what we had, but it's definitely a cleaner API. 

When it comes to names, what about changing `Configs` to `Instances`?

Reply:
> I think it's fine. It has a bit more boilerplate than what we had, but it's definitely a cleaner API.

Yes - I was aware of a bit more boilerplate, but I'm really convinced the API is cleaner.

> When it comes to names, what about changing Configs to Instances?

Fine with me.

Reply:
thanks @mm4tt , @wojtek-t . I have made the changes 
While trying to push the changes, i am facing an issue golint here:

https://github.com/kubernetes/perf-tests/commit/1d656288cdfe1feba7e0c114c418cbe5f49cea4d#diff-c137114a1d7f15af741f5213ca5c8ad8L30

This is from the latest commit above ^^

Getting this error:

> !!! 'golint' problems: 
/Users/s0r01ed/go/src/k8s.io/perf-tests/clusterloader2/pkg/util/ssh.go:36:1: comment on exported method GCloudSSHExecutor.Exec should be of the form "Exec ..."
error: failed to push some refs to 'https://github.com/vamossagar12/perf-tests.git'

Reply:
Yeah - seems like a bug. Would you be able to send a separate PR to fix it?

Reply:
Here you go: https://github.com/kubernetes/perf-tests/pull/1155

Reply:
That was merged - hopefully you should be able to update this PR now.

Reply:
yeah i was able to push it now.